File: mesa/demo/apps/simple_acl.rb
#!/usr/bin/env ruby

# Copyright (c) 2004-2020 Microchip Technology Inc. and its subsidiaries.
# SPDX-License-Identifier: MIT

require_relative 'libeasy/et'

$ts = get_test_setup("mesa_pc_b2b_2x")

$idx_tx = 0
$idx_rx = 1
$ace_id = 1

#---------- Configuration -----------------------------------------------------

test "conf" do
    # Set the default port action for the ingress port to be the drop action.
    # Any frame arriving on this port that doesn't match any ACE will be dropped.
    port = $ts.dut.port_list[$idx_tx]
    conf = $ts.dut.call("mesa_acl_port_conf_get", port)
    conf["action"]["port_action"] = "MESA_ACL_PORT_ACTION_FILTER"
    # empty port list = drop action
    conf["action"]["port_list"]   = ""
    $ts.dut.call("mesa_acl_port_conf_set", port, conf)

    # Add the correct rule to 
    # match : IPv4 + TCP + dport 80, ingress on port $idx_tx 
    # action : FILTER to $idx_rx (forward only to the rx port)
    ace = $ts.dut.call("mesa_ace_init", "MESA_ACE_TYPE_IPV4")
    ace["id"]        = $ace_id
    ace["port_list"] = port_idx_list_str([$idx_tx])

    ace["action"]["port_action"] = "MESA_ACL_PORT_ACTION_FILTER"
    ace["action"]["port_list"]   = port_idx_list_str([$idx_rx])

    # Match the transport proto to be TCP
    f = ace["frame"]["ipv4"]
    f["proto"]["value"] = 6
    f["proto"]["mask"]  = 0xff

    # Match the TCP dport to be 80
    f["dport"]["in_range"] = true
    f["dport"]["low"]      = 80
    f["dport"]["high"]     = 80

    $ts.dut.call("mesa_ace_add", 0, ace)
end

#---------- Frame testing -----------------------------------------------------

test "frame-io" do
    # Send IPv4 TCP frames with dport 70..90.
    (70..90).each do |dport|
        # Print the dport of the frame and the expected action  
        expected_rx = (dport == 80) ? [$idx_rx] : []
        t_i("TCP dport #{dport}, #{dport == 80 ? "forward to port #{$idx_rx}" : "drop"}")
        run_ef_tx_rx_cmd($ts, $idx_tx, expected_rx, "eth ipv4 tcp dport #{dport}")
    end
    
end

test_summary

File: mesa/demo/examples/simple-acl.c
// Copyright (c) 2004-2020 Microchip Technology Inc. and its subsidiaries.
// SPDX-License-Identifier: MIT

#include <unistd.h>
#include <stdio.h>
#include "cli.h"
#include "example.h"
#include "microchip/ethernet/switch/api.h"

#define ACE_ID_FWD  1 // Higher Priority: Forward TCP/80 to eport
#define ACE_ID_DROP 2 // Lower Priority: Drop everything else

static int acl_init(int argc, const char *argv[])
{
    mesa_port_no_t iport = ARGV_INT("iport", "Ingress Port");
    mesa_port_no_t eport = ARGV_INT("eport", "Egress Port / Port that receives TCP/80 frames");
    mesa_ace_t     ace;
    mesa_ace_frame_ipv4_t *ipv4 = &ace.frame.ipv4;

    EXAMPLE_BARRIER(argc);

    // ACE 1 (Higher Priority): forward to eport only IPv4 TCP dport 80
    RC(mesa_ace_init(NULL, MESA_ACE_TYPE_IPV4, &ace));
    ace.id = ACE_ID_FWD;
    mesa_port_list_set(&ace.port_list, iport, 1);
    ipv4->proto.value = 6;
    ipv4->proto.mask = 0xff;
    ipv4->dport.in_range = TRUE;
    ipv4->dport.low = 80;
    ipv4->dport.high = 80;
    ace.action.port_action = MESA_ACL_PORT_ACTION_REDIR;
    mesa_port_list_set(&ace.action.port_list, eport, 1);
    RC(mesa_ace_add(NULL, MESA_ACE_ID_LAST, &ace));

    // ACE 2 (Lower Priority): else drop all on iport
    RC(mesa_ace_init(NULL, MESA_ACE_TYPE_ANY, &ace));
    ace.id = ACE_ID_DROP;
    mesa_port_list_set(&ace.port_list, iport, 1);
    ace.action.port_action = MESA_ACL_PORT_ACTION_FILTER;
    RC(mesa_ace_add(NULL, MESA_ACE_ID_LAST, &ace));

    return 0;
}

static int acl_uninit(void)
{
    RC(mesa_ace_del(NULL, 1));
    RC(mesa_ace_del(NULL, 2));
    return 0;
}

static const char *help_txt =
    " Simple ACL example. \n"
    "\n"
    "A single ACE forwards IPv4 TCP frames with destination port 80 to the egress port.\n"
    "Any other frame will be dropped by default.\n";

static const char *acl_help(void) { return help_txt; }

EXAMPLE(simple_acl, acl_init, NULL, acl_uninit, acl_help);
